lib/ferrum/page/tracing.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+module Ferrum
+  class Page
+    class Tracing
+      INCLUDED_CATEGORIES = %w[
+        devtools.timeline
+        v8.execute
+        disabled-by-default-devtools.timeline
+        disabled-by-default-devtools.timeline.frame
+        toplevel
+        blink.console
+        blink.user_timing
+        latencyInfo
+        disabled-by-default-devtools.timeline.stack
+        disabled-by-default-v8.cpu_profiler
+        disabled-by-default-v8.cpu_profiler.hires
+      ].freeze
+      EXCLUDED_CATEGORIES = %w[
+        *
+      ].freeze
+
+      def initialize(client:)
+        self.client = client
+      end
+
+      def start(trace_params: {}, **options)
+        self.options = {
+          screenshots: false,
+          encoding: :binary,
+          **options
+        }
+        self.promise = Concurrent::Promises.resolvable_future
+        subscribe_on_tracing_event
+        inner_start(trace_params)
+      end
+
+      def stop
+        client.command("Tracing.end")
+        promise.value!
+      end
+
+      private
+
+      attr_accessor :client, :options, :promise
```

Comment:
Move it above `initialize`

Reply:
do you mean this?
```ruby
module Ferrum
  class Page
    class Tracing
      attr_accessor :client, :options, :promise

      def initialize(client:)
        self.client = client
      end
```
shall we say, it's encapsulation point because the order makes sense to keep methods `private`:
```ruby
irb(main):001:1* class Test
irb(main):002:1*   attr_accessor :string
irb(main):003:1*   
irb(main):004:1*   private
irb(main):005:1*   
irb(main):006:1*   attr_accessor :private_string
irb(main):007:0> end
=> [:private_string, :private_string=]
irb(main):008:0> Test.new.string = "public"
=> "public"
irb(main):009:0> Test.new.private_string = "public"
(irb):9:in `<main>': private method `private_string=' called for #<Test:0x0000558fe07472a8> (NoMethodError)
Did you mean?  private_methods
	from /home/mifrill/.rbenv/versions/3.0.2/lib/ruby/gems/3.0.0/gems/irb-1.3.5/exe/irb:11:in `<top (required)>'
	from /home/mifrill/.rbenv/versions/3.0.2/bin/irb:23:in `load'
	from /home/mifrill/.rbenv/versions/3.0.2/bin/irb:23:in `<main>'
```

Reply:
I don't see any reason to use private attr_accessor, most likely you will use instance variable instead.

Reply:
sure, no problem, it makes sense to be consistent with the current codebase style

Reply:
It just in general doesn't make sense, current codebase style is not the reason.

Reply:
speaking frankly, for me it's not a big deal, however, I think it makes sense to use always methods by `attr_`  instead `@` references due to things like possible extend or code-completion by IDE etc. By the way, it also makes sense to be explicit of what variables should be assignable or not by using `attr_writter` and it also makes sense for me to be strict about what of those variables should be available outside of the class by `private` order usage. Anyway, let's use `@` - no problem at all.

Reply:
They are mutually exclusive in my opinion. If you want to add additional logic to the method, you declare it explicitly with explicit access modifier, if not you can simply use instance variable and last you can use attr methods to declare public access to your state. So I have different opinion, you always use instance variables until you need to provide public access to your internal state. That’s my motto until proven wrong